### ISSUE TITLE
Uniformize TemporalPooler and TemporalMemory names

### ISSUE THREAD

**Issue by david-ragazzi** [numenta/nupic#1898]
Currently nupic has 2 types of temporal processors: TemporalPooler and TemporalMemory. Both are implemented in "pure python" and "hibrid python with C++" versions. However their names are not suggestive enough to someone correlates such implementations. I suggest we rename TemporalPoolers tastes to correspond to TemporalMemory name. In this case:
- `TP` (pure python TemporalPooler) become `temporal_pooler/TemporalPooler` to correlate to `temporal_memory.py/TemporalMemory`
- `TP10x2` (hybrid python TemporalPooler) become `fast_temporal_pooler.py/FastTemporalPooler` to correlate to `fast_temporal_memory.py/FastTemporalMemory`

Once serialization is done, this simple job could be

**Comment by scottpurdy** [comment]
Thanks for filing this David. We would probably want to move temporal memory implementations away from the name "temporal pooler" since they don't implement temporal pooling. This is a long-standing naming issue that we haven't been able to resolve in the code yet. That said, I'm not sure the best way to distinguish between the different implementations. I think a name that conveys their functional difference would be better than something like "old_temporal_memory" (implies outdated or worse) or similar.

**Comment by david-ragazzi** [comment]
> I think a name that conveys their functional difference would be better than something like "old_temporal_memory" (implies outdated or worse) or similar.

The problem is finding which name better describes such TP/TP10x2 _functional difference_.. Anyway, maybe such _functional difference_ could be simply moved into new temporal memory files/classes (not preserved as an old version) as @subutai suggest here:

>I suggest we just focus on creating a C++ version of temporal_memory.py, ensure it works really well and fast, and move everything to it. Then we can just delete Cells4, TP.py, and TP10X2.py This would remove all the naming confusion.
https://github.com/numenta/nupic.core/issues/351#issuecomment-75829073

In either case, if even so we reach the consensus that it is better we keep such functional difference in separate files, then I would suggest we use omnibus/generic names like `SpatialProcessor` and `TemporalProcessor` which could atenuate this and future confusions. I.e. not matter if a TP has pooling, memory sequence, whatever, it still would be a TemporalProcessor. To distinguish them, we could use (for example):
- SP => `spatial_processor_v1.py / SpatialProcessorV1`
- TP => `temporal_processor_v1.py / TemporalProcessorV1`
- TP10x2 => `fast_temporal_processor_v1.py / FastTemporalProcessorV1`
- TemporalMemory => `temporal_processor_v2.py / TemporalProcessorV2`
- FastTemporalMemory => `fast_temporal_processor_v2.py / FastTemporalProcessorV2`

(..we could use other convention than `V1` and `V2` to distinguish versions)

In summary, we could find what is the `TP/TP10x2` functional difference, if such functional difference can be moved into new (pure python and c++) `temporal_memory` implementations or then keep it in separated files as is currently (but using generic names with versioning).

**Comment by david-ragazzi** [comment]
> I think a name that conveys their functional difference would be better than something like "old_temporal_memory" (implies outdated or worse) or similar.

The problem is finding which name better describes such TP/TP10x2 _functional difference_.. Anyway, maybe such _functional difference_ could be simply moved into new temporal memory files/classes (not preserved as an "old" version) as @subutai suggest here:

>I suggest we just focus on creating a C++ version of temporal_memory.py, ensure it works really well and fast, and move everything to it. Then we can just delete Cells4, TP.py, and TP10X2.py This would remove all the naming confusion.
https://github.com/numenta/nupic.core/issues/351#issuecomment-75829073

In either case, if even so we reach the consensus that it is better we keep such functional difference in separate files, then I would suggest we use omnibus/general names like `SpatialProcessor` and `TemporalProcessor` which could atenuate this and future confusions. I.e. not matter if a TP has pooling, memory sequence, whatever, it still would be a TemporalProcessor. To distinguish them, we could use (for example):
- `SP` => `spatial_processor_v1.py / SpatialProcessor`
- `TP` => `temporal_processor_v1.py / TemporalProcessor`
- `TP10x2` => `fast_temporal_processor_v1.py / FastTemporalProcessor`
- `TemporalMemory` => `temporal_processor_v2.py / TemporalProcessor`
- `FastTemporalMemory` => `fast_temporal_processor_v2.py / FastTemporalProcessor`

When people want import different versions in a same script, they could use aliases:

``` 
from nupic.research.spatial_processor_v1 import SpatialProcessor
from nupic.research.temporal_processor_v1 import TemporalProcessor as TemporalProcessorV1
from nupic.research.temporal_processor_v2 import TemporalProcessor as TemporalProcessorV2
```

In case of you create another kind of temporal processor in the future, you simply would create a file called `temporal_processor_v3` and keep `TemporalProcessor` class name. (..we could use other convention than `V1` and `V2` to distinguish versions)

This convention has several advantages:
- It keeps history of implementations.
- Improve naming consistency between different implementations.
- Avoid future naming confusions.

In summary, we could find what is the `TP/TP10x2` functional difference, if such functional difference can be moved into new (pure python and c++) `temporal_memory` implementations or then keep it in separated files as already is currently (but using omnibus names with versioning).

**Comment by rcrowder** [comment]
Anything to make it easier for new people and drop confusing naming. Proposed changes seem fine for engineers, but may need closer mapping to biological processes in the future when more of the cortical layers are worked out.

At the moment Temporal Memory looks like it's acting at the cellular level (pyramidal). It maybe better to find a more appropriate name, to distinguish between non-cortex processing. If the Hippocampus and Medial Temporal Lobe have another method for memory storage, another name would be required there to separate it from the cortical memory processing (TM).

Whether the 'Temporal ' and 'Spatial' words remain undisturbed I'm unsure about. 'Pooler' seems appropriate to be retained. Although a closer naming to the actual functionality may be more appropriate.

**Comment by david-ragazzi** [comment]
> Proposed changes seem fine for engineers, but may need closer mapping to biological processes in the future when more of the cortical layers are worked out.

Because this, I am suggesting separate present and future implementations by versions but keeping `spatial` and `temporal` prefixes. In the end, although different implementations and purposes they still are processors of spatial and temporal properties.

**Comment by david-ragazzi** [comment]
Anyway, this discussion is quite old and it seems that is infinite because every time that we try using the `Specific Name` approach to naming such components, everyone have his own conventions (ML is a great example of these long discussions). What I'm suggesting is just we dissociate such specific names from such spatial and temporal processor class names and document their functional differences in their file descriptions and wiki. As far I know, thinking as engineer (and even as amateur neuroscientist), the objective of machine learning is extract the cognitive properties of brain and implement them in an abstract way (because this I'm suggesting broader names). Such convention is one step on this objective.

**Comment by scottpurdy** [comment]
@david-ragazzi - `v1`, `v2`, etc is fine with me. But the name is "temporal memory" and since we relatively recently settled on the naming I don't think we should reopen that discussion any time soon unless there is a real need.

**Comment by rhyolight** [comment]
> @david-ragazzi - v1, v2, etc is fine with me. But the name is "temporal memory" and since we relatively recently settled on the naming I don't think we should reopen that discussion any time soon unless there is a real need.

:+1:

**Comment by david-ragazzi** [comment]
> @david-ragazzi - v1, v2, etc is fine with me. But the name is "temporal memory" and since we relatively recently settled on the naming I don't think we should reopen that discussion any time soon unless there is a real need.

Ok boss.. Let's use `temporal_memory_v1` for `TP` and `TP10x2` and `temporal_memory_v2` for `TemporalMemory`. Once serialization is done, I'll work on this.

**Comment by chessweb01** [comment]
Let me use this thread to ask a couple of questions that I'm not able to 
answer satisfactorily for myself by studying the code:

1. What are the high-level differences between the temporal pooler 
(TP.py) and the temporal memory (temporal_memory.py) algorithms?
2. Are Jeff's new ideas about temporal pooling being implemented into 
temporal_memory.py or is that something else entirely?

Regards , RS


Am 02.03.2015 um 03:16 schrieb David Ragazzi:
